packages/bitcore-node/src/services/pruning.ts

```diff
@@ -112,144 +121,173 @@ export class PruningService {
-    });
-    logger.info(`Found ${count} invalid ${chain} ${network} txs`);
-    await new Promise((resolve, reject) => {
-      this.transactionModel.collection
-        .find({ chain, network, blockHeight: SpentHeightIndicators.conflicting })
-        .pipe(
-          new Transform({
-            objectMode: true,
-            transform: async (data: any, _, cb) => {
-              if (this.stopping) {
-                return cb(new Error('Stopping'));
-              }
-              const tx = data as ITransaction;
-              logger.info(`Invalidating ${tx.txid} outputs and dependent outputs`);
-              const outputGenerator = this.transactionModel.yieldRelatedOutputs(tx.txid);
-              let spentTxids = new Set<string>();
-              for await (const coin of outputGenerator) {
-                if (coin.mintHeight >= 0 || coin.spentHeight >= 0) {
-                  return cb(new Error(`Invalid coin! ${coin.mintTxid} `));
-                }
-                if (coin.spentTxid) {
-                  spentTxids.add(coin.spentTxid);
-                  if (spentTxids.size > DESCENDANT_LIMIT) {
-                    logger.warn(`${tx.txid} has too many decendants`);
-                    return cb();
-                  }
-                }
-              }
-              spentTxids.add(tx.txid);
-              const uniqueTxids = Array.from(spentTxids);
-              await this.clearInvalid(uniqueTxids);
-              logger.info(`Invalidated tx ${tx.txid} and ${spentTxids.size - 1} dependent txs`);
-              cb();
+    try {
+      if (this.runningInvalid) {
+        return;
+      }
+      this.runningInvalid = true;
+
+      if (Date.now() - this.lastRunTimeInvalid < INV_INTERVAL_MINS * ONE_MIN) {
+        return;
+      }
+
+      const count = await this.coinModel.collection.countDocuments({ chain, network, mintHeight: SpentHeightIndicators.pending });
+      logger.info(`Found ${count} pending ${chain}:${network} TXOs`);
+
+      // Note, realCount <= count since the coinStream returns coins dynamically and related coins are updated.
+      // The caveat is that new coins could be added to the stream while we are iterating over it.
+      let realCount = 0;
+      let invalidCount = 0;
+      const seen = new Set<string>();
+      const voutStream = this.coinModel.collection.find({ chain, network, mintHeight: SpentHeightIndicators.pending }).batchSize(1);
+      for await (const vout of voutStream) {
+        if (this.stopping) {
+          break;
+        }
+        realCount++;
+        if (seen.has(vout.mintTxid)) {
+          continue;
+        }
+        seen.add(vout.mintTxid);
+        const tx = await this.transactionModel.collection.findOne({ chain, network, txid: vout.mintTxid });
+        if (!tx) {
+          logger.error(`Coin ${vout.mintTxid} has no corresponding tx`);
+          continue;
+        }
+        if (tx.replacedByTxid) {
+          if (await this.invalidateTx(chain, network, tx)) {
```

Comment:
Using something that can update a DB directly in a conditional may lead to developer headaches. Consider assigning to 'didInvalidateTx' directly above this line & then checking that assignment in the conditional. Same on line 251